### ISSUE TITLE
Vertx Generator fails with 3.2.14 

### ISSUE THREAD

**Issue by matzew** [fabric8io/fabric8-maven-plugin#780]
### Description
Going from `3.2.8` to `3.2.14` causes simple vertx project to fail, when building (Can not extract generator)
### Info

running `mvn clean install -X` gives me this failiure (with the .14 version):
```
[WARNING] F8: Cannot access cluster for detecting mode: Failed to connect to /192.168.178.91:8443
[INFO] F8: Running in Kubernetes mode
[INFO] F8: Running generator vertx
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 1.882 s
[INFO] Finished at: 2017-01-15T13:32:45+01:00
[INFO] Final Memory: 25M/383M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.fabric8:fabric8-maven-plugin:3.2.14:resource (default) on project http-server: Execution default of goal io.fabric8:fabric8-maven-plugin:3.2.14:resource failed: Cannot extract generator: java.lang.NullPointerException -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal io.fabric8:fabric8-maven-plugin:3.2.14:resource (default) on project http-server: Execution default of goal io.fabric8:fabric8-maven-plugin:3.2.14:resource failed: Cannot extract generator: java.lang.NullPointerException
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default of goal io.fabric8:fabric8-maven-plugin:3.2.14:resource failed: Cannot extract generator: java.lang.NullPointerException
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:145)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
	... 20 more
Caused by: java.lang.IllegalArgumentException: Cannot extract generator: java.lang.NullPointerException
	at io.fabric8.maven.plugin.mojo.build.ResourceMojo$3.customizeConfig(ResourceMojo.java:620)
	at io.fabric8.maven.docker.config.ConfigHelper.resolveImages(ConfigHelper.java:51)
	at io.fabric8.maven.plugin.mojo.build.ResourceMojo.getResolvedImages(ResourceMojo.java:593)
	at io.fabric8.maven.plugin.mojo.build.ResourceMojo.executeInternal(ResourceMojo.java:225)
	at io.fabric8.maven.plugin.mojo.AbstractFabric8Mojo.execute(AbstractFabric8Mojo.java:74)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	... 21 more
Caused by: java.lang.NullPointerException
	at io.fabric8.maven.generator.api.support.AbstractPortsExtractor.extract(AbstractPortsExtractor.java:74)
	at io.fabric8.maven.generator.vertx.VertxGenerator.extractPorts(VertxGenerator.java:80)
	at io.fabric8.maven.generator.javaexec.JavaExecGenerator.customize(JavaExecGenerator.java:116)
	at io.fabric8.maven.plugin.generator.GeneratorManager.generate(GeneratorManager.java:62)
	at io.fabric8.maven.plugin.mojo.build.ResourceMojo$3.customizeConfig(ResourceMojo.java:618)
	... 26 more
[ERROR] 
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
```

* f-m-p version : 3.2.14
* Maven version (`mvn -v`) : 
```
Apache Maven 3.3.9 (NON-CANONICAL_2016-07-01T11:53:38Z_mockbuild; 2016-07-01T13:53:38+02:00)
Maven home: /usr/share/maven
Java version: 1.8.0_111, vendor: Oracle Corporation
Java home: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.111-3.b16.fc25.x86_64/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "4.8.8-300.fc25.x86_64", arch: "amd64", family: "unix"
```
* Sample project : `git@github.com:matzew/vertx-hello-world.git`  

The latest commit uses the 3.2.14 version of the f-m-p

**Comment by rhuss** [comment]
Thanks, its a regression indeed when no vert.x port configuration is given

@iocanel @kameshsampath @cescoffier I disabled now the addition of any port if not a port if configured either with the property `vertx.port` or within the configuration file pointed to by the v-m-p in its `<configuration><config>/path/to/config</config></configuration>`. I wonder whether in this case we still should add a default port like 8080 ?

**Comment by matzew** [comment]
@rhuss I am not using the `v-m-p`, should I?

Or what would be the exact workaround for the lacking port?

**Comment by rhuss** [comment]
You can use the system property (or maven project property) `vertx.config` pointing to a properties files which contains a property with "port" in its name, pointing to your port to expose.

E.g.;

```
webPort=8080
```

**Comment by rhuss** [comment]
I wonder whether we should still allow for a default port if no config if found. Currently, for the no-config case, no Service will be created. So I suggest to keep the behaviour of spring-boot and java-exec to use default ports, except when they are overwritten by configs.

Also for the `vertx.config` property I suggest to also evaluate it as a Maven property, not only a System property.

**Comment by rhuss** [comment]
If there are no objections I update #781 accordingly.

**Comment by cescoffier** [comment]
+1

Thanks @rhuss !

**Comment by cescoffier** [comment]
Actually, thinking about it, the approach to rely on the configuration file does not really match with Vert.x. We don't use file to configure the port and the new vertx-config provides lots of different way to configure the applications (including files of course, but many other ways too). Typically here, it may be a config map. 

I believe we should either use a project property (`vertx.port`) with a potential fallback to 8080 (that can be disabled if `vert.port` is set to empty, 0 or a negative value.

**Comment by rhuss** [comment]
That's a good idea, I'm going to implement it like this and default to 8080 which can be disabled.

**Comment by rhuss** [comment]
Just checked: This alread works with using the commonly used property `webPort` which is set by default to 8080. (property: `fabric8.generator.vertx.webPort`). If set to 0, its not added. This is the common behaviour of fat-jar generators.

**Comment by cescoffier** [comment]
@rhuss 0 only, what about negative numbers? Not something very important, but I saw people using -1 to disable HTTP ports (0 generally means a random free port).

**Comment by rhuss** [comment]
Good point, will add that, too.

**Comment by rhuss** [comment]
Is available now in 3.2.16
